src/tactic/transform_decl.lean

```diff
@@ -8,29 +8,24 @@ import meta.rb_map
 import tactic.core
 
 namespace tactic
+
 /-- `copy_attribute' attr_name src tgt p d_name` copy (user) attribute `attr_name` from
-   `src` to `tgt` if it is defined for `src`; and copy the parameter of the user attribute, in the
-   user attribute case. Make it persistent if `p` is `tt`; if `p` is `none`, the copied attribute is
-   made persistent iff it is persistent on `src`  -/
+   `src` to `tgt` if it is defined for `src`; unlike `copy_attribute` the primed version also copies
+   the parameter of the user attribute, in the user attribute case. Make it persistent if `p` is
+   `tt`; if `p` is `none`, the copied attribute is made persistent iff it is persistent on `src`  -/
 meta def copy_attribute' (attr_name : name) (src : name) (tgt : name) (p : option bool := none) :
 tactic unit :=
-try $ do
+do
   (p', prio) ← has_attribute attr_name src,
   let p := p.get_or_else p',
-get_decl tgt <|> fail!"unknown declaration {tgt}",
-s ← try_or_report_error (set_basic_attribute attr_name tgt p prio),
-sum.inr msg ← return s | skip,
-if msg =
-(format!"set_basic_attribute tactic failed, '{attr_name}' is not a basic attribute").to_string
-then do
-  user_attr_nm ← get_user_attribute_name attr_name,
-  user_attr_const ← mk_const user_attr_nm,
-  tac ← eval_pexpr (tactic unit)
-  ``(user_attribute.get_param %%user_attr_const %%src >>=
-      λ x, user_attribute.set %%user_attr_const %%tgt x %%p) <|>
-    (fail!"Cannot set attribute @[{attr_name}]."),
-  tac
-else fail msg
+  get_decl tgt <|> fail!"unknown declaration {tgt}",
+  set_basic_attribute attr_name tgt p prio <|>
+  (do get_user_attribute_name attr_name >>= (λ user_attr_nm : name, do
+    user_attr_const ← mk_const user_attr_nm,
+    tac ← eval_pexpr (tactic unit)
+    ``(user_attribute.get_param %%user_attr_const %%src >>=
+        λ x, user_attribute.set %%user_attr_const %%tgt x %%p %%prio),
+    tac))
```

Comment:
Could we maybe define a common generalization of `set_attribute` and `copy_attribute'` that takes as argument the `expr` that is fed into `eval_pexpr (tactic unit)`?

Btw, the reason thet `set_attribute` has the weird line 
```
if msg =  (format!"set_basic_attribute tactic failed, '{attr_name}' is not a basic attribute").to_string
```
is so that you get a reasonable error message when it is a basic attribute, but there is a failure giving that attribute to it (for example the "this is not a valid simp lemma"-error).

Reply:
This generalization sounds like a good idea, I've realised that we might need an extended version of this that applies the to_additive machinery to the parameter of the user attribute too, and maybe the generalization you mention will help with that too.

As for that line, yeah that makes sense, I basically only rewrote this so that I would understand it again. My plan is to instead check if a user_attribute exists first, and then if that fails just to run the basic attribute set tactic. I haven't tried it yet so maybe I'm missing something.